### ISSUE TITLE
Discussion for HW 1 Question 3

### ISSUE THREAD

**Issue by barbehenna** [uofi-stat430/fa20-disc#6]
Please use this issue for questions on homework 1, question 3.

**Comment by PeichenLi** [comment]
Hi,

I can't pass the test of Exits (Bad Continent) while others were fine.  I used "[[ $# -eq 2 ]]" to check the error condition. Did I miss any other conditions for returning an error code?

**Comment by PeichenLi** [comment]
Hi,

How to check if the continent exists? Thanks.

**Comment by barbehenna** [comment]
Each continent is stored is its own directory, for example Europe is a directory that contains csv files for specific European counties. This means that to check if the continent exists, you need to check for that the directory exists.

**Comment by eddelbuettel** [comment]
And that directory (by continent) is _identical_ to what is both the downloadable data set (in a tar.gz) and the data set git repo. So you can practice this on your machine, or on RStudio Cloud, or ...  in the exact same way you will find in CBTF and PL.

**Comment by bowend5** [comment]
Can anyone tell me how do I error out? I tried set -exu but it seems not to be the one. Also how do I exit with an exit code?

**Comment by PeichenLi** [comment]
After satisfying the error out condition, imply put exit -1

**Comment by bowend5** [comment]
Can anyone help me find what is wrong in my code? All the other tests are passed but I couldn't find what is wrong about the "Error, Bad continent" test. 

if [ $# -eq 2 ] || [ -d "/files_unfiltered/$1" ]
then
    wc -m $1/*csv|head -n -1|sort -n -r| head -$2| tail -1|tr -dc '0-9'
else
    exit -1
fi

**Comment by eddelbuettel** [comment]
I would suggest you simplify and test all your assumptions about 
- correct arguments
- correct directory
- correct counts

and so on.

**Comment by hanj2** [comment]
hi, 
I'm confused about 'n' and 'continent' here.
should we write our command into a function such as foo() and then call this function with arguments?
or should it only be some command with $1 and $2? then how should it be called with arguments?

**Comment by barbehenna** [comment]
Either approach is fine and will work. I suspect you will find it easier to simply write a script that uses $1 and $2. The hint tells you how the script will be called with arguments.

**Comment by hanj2** [comment]
Thank you! now I use  
> if [-d "$1"]; then ] after changing into the required dir
to check if a continent(as the first arg) exist as a sub dir
However, I get the same "continent does not exist" error message no matter it exists or not. 
Is there anything else I need to check?

**Comment by barbehenna** [comment]
"continent does not exist" does not sound like an error message that is coming from a shell command. Are you writing that statement? If so, it's not very helpful for us to help. 

It sounds like your error may be due to how you're handling the directories and `continent` input. The first input to the script is a continent, like "Americas" or "Europe", just like how in part 1, `continent` is set to "Europe". Additionally, make sure you're running your script in the correct location because if you're not, then the directories will not exist. For example, if you're in the `files_unfiltered` directory the directory `files_unfiltered/Europe` will not exist, but the directory `Europe` will exist.

**Comment by JingyingLuo** [comment]
Dear professor,
I am wondering if it is possible that the question 3 have some error in default setting now?
I submitted answer on weekend and it give me full credit, but when I practice again which I checked is exactly same as my previous answer it shows only 60% correct

**Comment by eddelbuettel** [comment]
Hard to tell -- we didn't change the exam setup, and the PrairieLearn team is experienced and does not generally break this either. 

It also "works for me" when I tried just now, both with a complete and a partial answer.

**Comment by JingyingLuo** [comment]
I think my friends and I faced the same problem with submitting the old answer but get only partial credit.

**Comment by JingyingLuo** [comment]
I think I figured out whats going on. 
when printing out the number of lines we probably need use cut -d' ' -f2 
which is correct in question 2
but here in question 3, I have to change f2 to f3 to get the correct answer.( but when I try to run in R cloud terminal, -f3 will output the file name instead of the number of line.
Why will this happened? I am confused now.

**Comment by barbehenna** [comment]
This is a tricky subtly in the question. If you use `wc` to get the counts you get something that looks like:
```
     596 Albania.csv
    2643 Austria.csv
    2653 Belgium.csv
     779 Bosnia and Herzegovina.csv
```
You'll notice that `wc` has lined up the right hand side of the numbers, this means that there may be different numbers of spaces in front of each number. This means that depending on the row you look at, when you run `cut -d" " -f2` you will either get a blank "column" or a number because `cut` is counting the number of spaces. The difference you're seeing between over the weekend and today may be explained by the autograder randomly picking a different `n` and `continent`.

**Comment by eddelbuettel** [comment]
I just consulted with PrarieLearn stuff and there is no issue.

You follow-up makes it clearer that the problem is your.  When we wrote this question (and the preceding one) we did reuse code.  From your partial description I cannot tell -- this would be a good topic for a follow-up in an office hour.  

But glad to know you're squared now.   (And small Markdown tip:  use backtick to typeset code:  `cut -f 2`.

**Comment by eddelbuettel** [comment]
Spot on, @barbehenna .  I didn't think hard enough about our own setup,

**Comment by JingyingLuo** [comment]
@barbehenna Thanks for your explanation. If cut command is not a general way to deal with the different spaces, is there any other command we could use to avoid the differences happen?
@eddelbuettel Thank you professor for your help! I will definitely follow up into next office hour.

**Comment by eddelbuettel** [comment]
There is.  "The next set of cool commands" are `sed` (a "stream editor") and the `awk` (after inventors Aho, Weinberger, Kernighan [yes the same Kernighan who co-invented C and Go]) but we didn't have time to cover that.  One super-useful feature of `awk` is that it tokenizes input _irrespective of whitespaces_ and processes line by line with the i-th token in $i.  So where:

```
$ echo "a b c d    e fff   g" | awk '{print $2, $6}'
b fff
$ 
```

you always get the second and sixth token no matter the whitespace.  But again, this goes a little beyond the course where we only have time for simpler things.

That said: in another issue ticket somebody already mentioned the `man` command, and I mention the excellent (shorter) "tldr" documentation.   If you look closely at the features of `cut` maybe you will find another way too...

**Comment by helphelpme** [comment]
It seems like prarielearn is not grading the problems.

**Comment by eddelbuettel** [comment]
Agreed. I escalated it at the PL Slack. I am confident someone will look into it shortly.

Already confirmed.  Time for expected fix unclear.

**Comment by eddelbuettel** [comment]
And taken care of.  From Professor West who is the main force behind all this:

![image](https://user-images.githubusercontent.com/673121/92670611-017d4280-f2da-11ea-9832-537311ff31ef.png)
